File: C#/Trivia/Trivia/Game.cs
using System;
using System.Collections.Generic;
using Trivia.Questions;

namespace Trivia
{
    public partial class Game : IGame<Game>
    {
        private readonly bool _useTechnoInsteadOfRock;
        private readonly List<Player> _players = new ();
        
        private readonly bool[] _inPenaltyBox = new bool[Configuration.NombreMaximalJoueurs + 1];

        private readonly QuestionsDeck _questions;

        private int _currentPlayer;
        private bool _isGettingOutOfPenaltyBox;

        public Game(bool useTechnoInsteadOfRock)
        {
            _useTechnoInsteadOfRock = useTechnoInsteadOfRock;
            _questions = new QuestionsDeck("Pop", "Science", "Sports", useTechnoInsteadOfRock ? "Techno" : "Rock");
        }

        // Constructeur copiant la partie en éliminant un joueur
        private Game(Game copied, Player playerToRemove)
        {
            var playerToRemoveId = copied._players.IndexOf(playerToRemove);

            _currentPlayer = copied._currentPlayer;
            if(_currentPlayer == playerToRemoveId) IncrementCurrentPlayer();

            _isGettingOutOfPenaltyBox = copied._isGettingOutOfPenaltyBox;

            for (var index = 0; index <= Configuration.NombreMaximalJoueurs; index++)
            {
                if(index == playerToRemoveId) continue;

                _inPenaltyBox[index] = copied._inPenaltyBox[index];
                if(copied._players.Count > index) _players.Add(copied._players[index]);
            }

            _questions = copied._questions.Save().Restore();
            _useTechnoInsteadOfRock = copied._useTechnoInsteadOfRock;
        }

        public void Add(string playerName)
        {
            _players.Add(new Player(playerName));
            _inPenaltyBox[NumberOfPlayers] = false;

            Console.WriteLine(playerName + " was added");
            Console.WriteLine("They are player number " + _players.Count);
        }

        public int NumberOfPlayers => _players.Count;
        private Player CurrentPlayer => _players[_currentPlayer % (NumberOfPlayers - 1)];

        /// <inheritdoc />
        public IGame<Game> GameWithoutAPlayer(Player playerToRemove) => new Game(this, playerToRemove);

        public void Roll(int roll)
        {
            Console.WriteLine(CurrentPlayer + " is the current player");
            Console.WriteLine("They have rolled a " + roll);

            if (_inPenaltyBox[_currentPlayer])
            {
                if (roll % 2 != 0)
                {
                    _isGettingOutOfPenaltyBox = true;

                    Console.WriteLine(CurrentPlayer + " is getting out of the penalty box");
                    CurrentPlayer.Move(roll);

                    Console.WriteLine(CurrentPlayer
                                      + "'s new location is "
                                      + CurrentPlayer.Place);
                    Console.WriteLine("The category is " + CurrentCategory());
                    AskQuestion();
                }
                else
                {
                    Console.WriteLine(CurrentPlayer + " is not getting out of the penalty box");
                    _isGettingOutOfPenaltyBox = false;
                }
            }
            else
            {
                CurrentPlayer.Move(roll);

                Console.WriteLine(CurrentPlayer
                                  + "'s new location is "
                                  + CurrentPlayer.Place);
                Console.WriteLine("The category is " + CurrentCategory());
                AskQuestion();
            }
        }

        private void AskQuestion()
        {
            Console.WriteLine(_questions.NextQuestionForCategory(CurrentCategory()));
        }

        private string CurrentCategory()
        {
            switch (CurrentPlayer.Place)
            {
                case 0:
                case 4:
                case 8:
                    return "Pop";
                case 1:
                case 5:
                case 9:
                    return "Science";
                case 2:
                case 6:
                case 10:
                    return "Sports";
                default:
                    return _useTechnoInsteadOfRock ? "Techno" : "Rock";
            }
        }

        private void IncrementCurrentPlayer()
        {
            _currentPlayer++;
            if (_currentPlayer == _players.Count) _currentPlayer = 0;
        }

        public Player? WasCorrectlyAnswered()
        {

            if (_inPenaltyBox[_currentPlayer])
            {
                if (_isGettingOutOfPenaltyBox)
                {
                    Console.WriteLine("Answer was correct!!!!");
                    CurrentPlayer.AddOneGold();
                    Console.WriteLine(CurrentPlayer
                                      + " now has "
                                      + CurrentPlayer.Purse
                                      + " Gold Coins.");

                    var winner = DidPlayerWin();
                    IncrementCurrentPlayer();
                    return winner;
                }
                else
                {
                    IncrementCurrentPlayer();
                    return null;
                }
            }
            else
            {
                Console.WriteLine("Answer was correct!!!!");
                CurrentPlayer.AddOneGold();
                Console.WriteLine(CurrentPlayer
                                  + " now has "
                                  + CurrentPlayer.Purse
                                  + " Gold Coins.");

                var winner = DidPlayerWin();
                IncrementCurrentPlayer();

                return winner;
            }
        }

        public Player? WrongAnswer()
        {
            Console.WriteLine("Question was incorrectly answered");
            Console.WriteLine(CurrentPlayer + " was sent to the penalty box");
            _inPenaltyBox[_currentPlayer] = true;

            IncrementCurrentPlayer();
            return null;
        }

        private Player? DidPlayerWin()
        {
            return CurrentPlayer.Purse == 6 ? CurrentPlayer : default;
        }
    }

}

File: kotlin/src/main/kotlin/com/adaptionsoft/games/uglytrivia/Game.kt
package com.adaptionsoft.games.uglytrivia

import com.adaptionsoft.games.uglytrivia.console.IConsole
import com.adaptionsoft.games.uglytrivia.console.SystemConsole
import com.adaptionsoft.games.uglytrivia.errors.MinimalGoldRequiredNotEnoughError
import com.adaptionsoft.games.uglytrivia.errors.PlayersNumberError
import java.lang.StringBuilder
import java.util.*
import kotlin.NoSuchElementException
import kotlin.collections.ArrayList

class Game(
    private val console: IConsole = SystemConsole(),
    private val minimalGoldRequired: Int = 6,
    private val shouldReplaceRockByTechno: Boolean = false,
    private val shouldUseExpansionPack: Boolean = false
) {
    private var players = arrayListOf<Player>()

    private var popQuestions = LinkedList<String>()
    private var scienceQuestions = LinkedList<String>()
    private var sportsQuestions = LinkedList<String>()
    private var rockQuestions = LinkedList<String>()
    private var technoQuestions = LinkedList<String>()
    private var rapQuestions = LinkedList<String>()
    private var philoQuestions = LinkedList<String>()
    private var litteratureQuestions = LinkedList<String>()
    private var geoQuestions = LinkedList<String>()

    private var currentPlayerIndex = 0
    private var isGettingOutOfPenaltyBox: Boolean = false

    private var numberOfQuestions = 0

    private lateinit var currentCategory: CAT
    private var isChoosingCategory = false

    enum class CAT {
        POP, SCIENCE, SPORT, ROCK, TECHNO, RAP, PHILO, LITTERATURE, GEO
    }

    init {
        for (i in 0..49) {
            popQuestions.addLast("Pop Question $i")
            scienceQuestions.addLast("Science Question $i")
            sportsQuestions.addLast("Sports Question $i")
            if (shouldReplaceRockByTechno) {
                technoQuestions.addLast("Techno Question $i")
            } else {
                rockQuestions.addLast("Rock Question $i")
            }
        }
    }

    fun getParameters(): String {
        val builder = StringBuilder()
        builder.append("minimalGoldRequired: $minimalGoldRequired")
        builder.append("shouldUseExpansionPack: $shouldUseExpansionPack")
        builder.append("shouldReplaceRockByTechno: $shouldReplaceRockByTechno")
        builder.append("players: ${players.map { it.name }}")
        return builder.toString()
    }

    fun add(newPlayers: ArrayList<Player>) {
        players = newPlayers
        players.forEachIndexed { index, player ->
            console.println("${player.name} was added")
            console.println("They are player number ${index + 1}")
        }
        throwIfNotPlayable()
    }

    fun roll(roll: Int) {
        console.println("${players[currentPlayerIndex].name} is the current player")
        console.println("They have rolled a $roll")

        if (players[currentPlayerIndex].isInPenaltyBox) {
            if (roll % 2 != 0) {
                isGettingOutOfPenaltyBox = true

                players[currentPlayerIndex].isInPenaltyBox = false
                console.println("${players[currentPlayerIndex].name} is getting out of the penalty box")
                players[currentPlayerIndex].place = players[currentPlayerIndex].place + roll
                if (players[currentPlayerIndex].place > 11) players[currentPlayerIndex].place = players[currentPlayerIndex].place - 12

                console.println(players[currentPlayerIndex].name
                        + "'s new location is "
                        + players[currentPlayerIndex].place)
                if (isChoosingCategory) {
                    console.println("The category is $currentCategory")
                    isChoosingCategory = false
                } else {
                    console.println("The category is " + currentCategory())
                }
                askQuestion()
            } else {
                console.println("${players[currentPlayerIndex].name} is not getting out of the penalty box")
                isGettingOutOfPenaltyBox = false
            }
        } else {
            players[currentPlayerIndex].place = players[currentPlayerIndex].place + roll
            if (players[currentPlayerIndex].place > 11) players[currentPlayerIndex].place = players[currentPlayerIndex].place - 12

            console.println(players[currentPlayerIndex].name
                    + "'s new location is "
                    + players[currentPlayerIndex].place)
            if (isChoosingCategory) {
                console.println("The category is $currentCategory")
                isChoosingCategory = false
            } else {
                console.println("The category is " + currentCategory())
            }
            askQuestion()
        }

    }

    fun wasCorrectlyAnswered(): Boolean {
        try {
            // Check if player has been removed because doesn't want to answer
            players[currentPlayerIndex]
        } catch (error: IndexOutOfBoundsException) {
            incrementCurrentPlayerIndex()
            if (players.size == 0) {
                return false
            }
            return true
        }
        if (players[currentPlayerIndex].hasJoker) {
            players[currentPlayerIndex].hasJoker = false
            incrementCurrentPlayerIndex()
            return true
        }
        if (players[currentPlayerIndex].isInPenaltyBox) {
            if (isGettingOutOfPenaltyBox) {
                console.println("Answer was correct!!!!")
                players[currentPlayerIndex].correctAnswersStrike++
                players[currentPlayerIndex].purses += players[currentPlayerIndex].correctAnswersStrike
                console.println(players[currentPlayerIndex].name
                        + " now has "
                        + players[currentPlayerIndex].purses
                        + " Gold Coins.")

                val winner = didPlayerNotWin()
                incrementCurrentPlayerIndex()

                return winner
            } else {
                incrementCurrentPlayerIndex()
                return true
            }
        } else {
            console.println("Answer was correct!!!!")
            players[currentPlayerIndex].correctAnswersStrike++
            players[currentPlayerIndex].purses += players[currentPlayerIndex].correctAnswersStrike
            console.println(players[currentPlayerIndex].name
                    + " now has "
                    + players[currentPlayerIndex].purses
                    + " Gold Coins.")

            val winner = didPlayerNotWin()
            currentPlayerIndex++
            if (currentPlayerIndex == players.size) currentPlayerIndex = 0

            return winner
        }
    }

    fun wrongAnswer(): Boolean {
        try {
            // Check if player has been removed because doesn't want to answer
            players[currentPlayerIndex]
        } catch (error: IndexOutOfBoundsException) {
            if (players.size == 0) {
                return false
            }
            incrementCurrentPlayerIndex()
            return true
        }
        if (players[currentPlayerIndex].hasJoker) {
            players[currentPlayerIndex].hasJoker = false
            incrementCurrentPlayerIndex()
            return true
        }
        console.println("Question was incorrectly answered")
        console.println(players[currentPlayerIndex].name + " was sent to the penalty box")
        players[currentPlayerIndex].correctAnswersStrike = 0
        players[currentPlayerIndex].isInPenaltyBox = true
        chooseNextCategory()

        incrementCurrentPlayerIndex()
        return true
    }

    fun chooseNextCategory() {
        isChoosingCategory = true
        /*console.println(players[currentPlayerIndex].name + " choosing category for next one : ")
        var userCATChoice = readln()
        while (!CAT.values().map { it.name }.contains(userCATChoice)) {
            console.println("Please choose between : ${CAT.values().map { it.name }}")
            userCATChoice = readln()
        }
        currentCategory = CAT.valueOf(userCATChoice)*/
        currentCategory = CAT.POP
    }

    private fun askQuestion() {
        if (players[currentPlayerIndex].hasJoker) {
            console.println("${players[currentPlayerIndex].name} uses his joker")
        } else if (players[currentPlayerIndex].doesNotWantToAnswer) {
            console.println("${players[currentPlayerIndex].name} doesn't want to answer and left the game")
            players.removeAt(currentPlayerIndex)
        } else {
            try {
                if (!isChoosingCategory) {
                    currentCategory()
                }
                when (currentCategory) {
                    CAT.POP -> console.println(popQuestions.removeFirst().toString())
                    CAT.SCIENCE -> console.println(scienceQuestions.removeFirst().toString())
                    CAT.SPORT -> console.println(sportsQuestions.removeFirst().toString())
                    CAT.ROCK -> console.println(rockQuestions.removeFirst().toString())
                    CAT.TECHNO -> console.println(technoQuestions.removeFirst().toString())
                    else -> {}
                }
            } catch (exception: NoSuchElementException) {
                fillQuestions()
                askQuestion()
            }
        }
    }

    private fun incrementCurrentPlayerIndex() {
        currentPlayerIndex++
        if (currentPlayerIndex >= players.size) {
            currentPlayerIndex = 0
        }
    }

    private fun currentCategory(): CAT {
        currentCategory = when (players[currentPlayerIndex].place) {
            0, 4, 8 -> CAT.POP
            1, 5, 9 -> CAT.SCIENCE
            2, 6, 10 -> CAT.SPORT
            else -> if (shouldReplaceRockByTechno) CAT.TECHNO else CAT.ROCK
        }
        return currentCategory
    }

    private fun throwIfNotPlayable() {
        if (players.size !in 2..6) throw PlayersNumberError()
        if (minimalGoldRequired < 6) throw MinimalGoldRequiredNotEnoughError()

    }

    private fun didPlayerNotWin(): Boolean = players[currentPlayerIndex].purses < minimalGoldRequired

    private fun fillQuestions() {
        for (i in numberOfQuestions..numberOfQuestions + 49) {
            popQuestions.addLast("Pop Question $i")
            scienceQuestions.addLast("Science Question $i")
            sportsQuestions.addLast("Sports Question $i")
            if (shouldReplaceRockByTechno) {
                technoQuestions.addLast("Techno Question $i")
            } else {
                rockQuestions.addLast("Rock Question $i")
            }
            if (shouldUseExpansionPack) {
                rapQuestions.addLast("Rap Question $i")
                philoQuestions.addLast("Philo Question $i")
                litteratureQuestions.addLast("Litterature Question $i")
                geoQuestions.addLast("Geo Question $i")
            }
        }
        numberOfQuestions += 49
    }
}